File: Trab TADSS/java/src/services/SimuladorFila.java
package src.services;

import src.models.Cliente;
import src.models.ConfiguracaoSimulacao;
import src.models.ResultadoSimulacao;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class SimuladorFila {
    private final ConfiguracaoSimulacao config;

    public SimuladorFila(ConfiguracaoSimulacao config) {
        this.config = config;
    }

    public ResultadoSimulacao executar(int quantidadeAtendentes, long seed) {
        Random random = new Random(seed);
        List<Cliente> clientes = gerarClientes(random);

        double[] disponibilidadeAtendentes = new double[quantidadeAtendentes];

        double somaEspera = 0.0;
        double maxEspera = 0.0;

        double somaAtendimento = 0.0;
        double maxAtendimento = 0.0;

        double somaLeadTime = 0.0;

        int concluidosAte13h = 0;
        int clientesDentroMeta = 0;

        for (Cliente cliente : clientes) {
            int indiceAtendente = encontrarAtendenteMaisLivre(disponibilidadeAtendentes);

            double inicioAtendimento = Math.max(cliente.instanteChegada(), disponibilidadeAtendentes[indiceAtendente]);
            double tempoEspera = inicioAtendimento - cliente.instanteChegada();
            double fimAtendimento = inicioAtendimento + cliente.tempoAtendimento();
            double leadTime = fimAtendimento - cliente.instanteChegada();

            disponibilidadeAtendentes[indiceAtendente] = fimAtendimento;

            somaEspera += tempoEspera;
            if (tempoEspera > maxEspera) {
                maxEspera = tempoEspera;
            }

            somaAtendimento += cliente.tempoAtendimento();
            if (cliente.tempoAtendimento() > maxAtendimento) {
                maxAtendimento = cliente.tempoAtendimento();
            }

            somaLeadTime += leadTime;

            if (fimAtendimento <= config.getDuracaoJanelaSegundos()) {
                concluidosAte13h++;
            }

            if (tempoEspera <= config.getMetaEsperaSegundos()) {
                clientesDentroMeta++;
            }
        }

        int totalClientes = clientes.size();

        ResultadoSimulacao resultado = new ResultadoSimulacao();
        resultado.setAtendentes(quantidadeAtendentes);
        resultado.setClientesGerados(totalClientes);
        resultado.setClientesConcluidosAte13h(concluidosAte13h);
        resultado.setTempoMedioEsperaFila(totalClientes == 0 ? 0 : somaEspera / totalClientes);
        resultado.setTempoMaximoEsperaFila(maxEspera);
        resultado.setTempoMedioAtendimento(totalClientes == 0 ? 0 : somaAtendimento / totalClientes);
        resultado.setTempoMaximoAtendimento(maxAtendimento);
        resultado.setLeadTimeMedio(totalClientes == 0 ? 0 : somaLeadTime / totalClientes);
        resultado.setPercentualClientesDentroMeta(totalClientes == 0 ? 0 : (clientesDentroMeta * 100.0 / totalClientes));

        return resultado;
    }

    private List<Cliente> gerarClientes(Random random) {
        List<Cliente> clientes = new ArrayList<>();

        double instanteAtual = 0.0;
        int id = 1;

        while (true) {
            int intervaloChegada = random.nextInt(
                    config.getChegadaMaxSegundos() - config.getChegadaMinSegundos() + 1
            ) + config.getChegadaMinSegundos();

            instanteAtual += intervaloChegada;

            if (instanteAtual > config.getDuracaoJanelaSegundos()) {
                break;
            }

            int tempoAtendimento = random.nextInt(
                    config.getAtendimentoMaxSegundos() - config.getAtendimentoMinSegundos() + 1
            ) + config.getAtendimentoMinSegundos();

            clientes.add(new Cliente(id++, instanteAtual, tempoAtendimento));
        }

        return clientes;
    }

    private int encontrarAtendenteMaisLivre(double[] disponibilidadeAtendentes) {
        int indiceMenor = 0;

        for (int i = 1; i < disponibilidadeAtendentes.length; i++) {
            if (disponibilidadeAtendentes[i] < disponibilidadeAtendentes[indiceMenor]) {
                indiceMenor = i;
            }
        }

        return indiceMenor;
    }
}

File: Trab TADSS/cs/Services/SimuladorFila.cs
using SimuladorFilas.Models;

namespace SimuladorFilas.Services
{
    public class SimuladorFila
    {
        private readonly ConfiguracaoSimulacao _config;

        public SimuladorFila(ConfiguracaoSimulacao config)
        {
            _config = config;
        }

        public ResultadoSimulacao Executar(int quantidadeAtendentes, int seed)
        {
            var random = new Random(seed);
            var clientes = GerarClientes(random);

            double[] disponibilidadeAtendentes = new double[quantidadeAtendentes];

            double somaEspera = 0;
            double maxEspera = 0;

            double somaAtendimento = 0;
            double maxAtendimento = 0;

            double somaLeadTime = 0;

            int concluidosAte13h = 0;
            int clientesDentroMeta = 0;

            foreach (var cliente in clientes)
            {
                int indiceAtendente = EncontrarAtendenteMaisLivre(disponibilidadeAtendentes);

                double inicioAtendimento = Math.Max(cliente.InstanteChegada, disponibilidadeAtendentes[indiceAtendente]);
                double tempoEspera = inicioAtendimento - cliente.InstanteChegada;
                double fimAtendimento = inicioAtendimento + cliente.TempoAtendimento;
                double leadTime = fimAtendimento - cliente.InstanteChegada;

                disponibilidadeAtendentes[indiceAtendente] = fimAtendimento;

                somaEspera += tempoEspera;
                if (tempoEspera > maxEspera)
                    maxEspera = tempoEspera;

                somaAtendimento += cliente.TempoAtendimento;
                if (cliente.TempoAtendimento > maxAtendimento)
                    maxAtendimento = cliente.TempoAtendimento;

                somaLeadTime += leadTime;

                if (fimAtendimento <= _config.DuracaoJanelaSegundos)
                    concluidosAte13h++;

                if (tempoEspera <= _config.MetaEsperaSegundos)
                    clientesDentroMeta++;
            }

            int totalClientes = clientes.Count;

            return new ResultadoSimulacao
            {
                Atendentes = quantidadeAtendentes,
                ClientesGerados = totalClientes,
                ClientesConcluidosAte13h = concluidosAte13h,
                TempoMedioEsperaFila = totalClientes == 0 ? 0 : somaEspera / totalClientes,
                TempoMaximoEsperaFila = maxEspera,
                TempoMedioAtendimento = totalClientes == 0 ? 0 : somaAtendimento / totalClientes,
                TempoMaximoAtendimento = maxAtendimento,
                LeadTimeMedio = totalClientes == 0 ? 0 : somaLeadTime / totalClientes,
                PercentualClientesDentroMeta = totalClientes == 0 ? 0 : (clientesDentroMeta * 100.0 / totalClientes)
            };
        }

        private List<Cliente> GerarClientes(Random random)
        {
            var clientes = new List<Cliente>();

            double instanteAtual = 0;
            int id = 1;

            while (true)
            {
                int intervaloChegada = random.Next(_config.ChegadaMinSegundos, _config.ChegadaMaxSegundos + 1);
                instanteAtual += intervaloChegada;

                if (instanteAtual > _config.DuracaoJanelaSegundos)
                    break;

                int tempoAtendimento = random.Next(_config.AtendimentoMinSegundos, _config.AtendimentoMaxSegundos + 1);

                clientes.Add(new Cliente(id++, instanteAtual, tempoAtendimento));
            }

            return clientes;
        }

        private int EncontrarAtendenteMaisLivre(double[] disponibilidadeAtendentes)
        {
            int indiceMenor = 0;

            for (int i = 1; i < disponibilidadeAtendentes.Length; i++)
            {
                if (disponibilidadeAtendentes[i] < disponibilidadeAtendentes[indiceMenor])
                    indiceMenor = i;
            }

            return indiceMenor;
        }
    }
}